crates/autopilot/src/limit_order_quoter.rs

```diff
@@ -131,4 +86,60 @@ impl LimitOrderQuoter {
         Metrics::on_failed(failed_orders.try_into().unwrap());
         Ok(())
     }
+
+    async fn update_surplus_fee(&self, order: &Order) -> Result<()> {
+        let quote = self
+            .quoter
+            .calculate_quote(QuoteParameters {
+                sell_token: order.data.sell_token,
+                buy_token: order.data.buy_token,
+                side: match order.data.kind {
+                    OrderKind::Buy => OrderQuoteSide::Buy {
+                        buy_amount_after_fee: order.data.buy_amount,
+                    },
+                    OrderKind::Sell => OrderQuoteSide::Sell {
+                        sell_amount: SellAmount::BeforeFee {
+                            value: order.data.sell_amount + order.data.fee_amount,
+                        },
+                    },
+                },
+                from: order.metadata.owner,
+                app_data: order.data.app_data,
+                signing_scheme: match &order.signature {
+                    Signature::Eip712(_) => QuoteSigningScheme::Eip712,
+                    Signature::EthSign(_) => QuoteSigningScheme::EthSign,
+                    Signature::Eip1271(signature) => {
+                        // TODO What about SignatureConfiguration like in order_validation.rs?
+                        let additional_gas = self
+                            .signature_validator
+                            .validate_signature_and_get_additional_gas(SignatureCheck {
+                                signer: order.metadata.owner,
+                                hash: hashed_eip712_message(
+                                    &self.domain_separator,
+                                    &order.data.hash_struct(),
+                                ),
+                                signature: signature.to_owned(),
+                            })
+                            .await?;
```

Comment:
`update_surplus_fee()` is still quite long. Maybe the additional fee computation could also be moved into a separate function?

Reply:
Improved error handling is of course also a nice added benefit but code is mostly read so collapsing a lot of verbose code into a single well named function call reduces mental overhead for the reader a lot IMO.

`let additional_gas = self.compute_signature_validation_gas(order)?` would be equally correct but way easier to parse. However, if you prefer having that code in-line this is not a hill to die on for me.